packages/core/saved-objects/core-saved-objects-migration-server-internal/src/kibana_migrator_utils.ts

```diff
@@ -15,21 +15,35 @@ import { TypeStatus, type TypeStatusDetails } from './kibana_migrator_constants'
 
 // even though this utility class is present in @kbn/kibana-utils-plugin, we can't easily import it from Core
 // aka. one does not simply reuse code
-export class Defer<T> {
-  public resolve!: (data: T) => void;
+class Defer<T> {
+  public resolve!: (data?: T) => void;
   public reject!: (error: any) => void;
   public promise: Promise<any> = new Promise<any>((resolve, reject) => {
     (this as any).resolve = resolve;
     (this as any).reject = reject;
   });
 }
 
-export const defer = () => new Defer<void>();
+export type WaitGroup<T> = Defer<T>;
 
-export function createMultiPromiseDefer(indices: string[]): Record<string, Defer<void>> {
-  const defers: Array<Defer<void>> = indices.map(defer);
-  const all = Promise.all(defers.map(({ promise }) => promise));
-  return indices.reduce<Record<string, Defer<any>>>((acc, indexName, i) => {
+export function waitGroup<T>(): WaitGroup<T> {
+  return new Defer<T>();
+}
+
+export function createWaitGroupMap<T, U>(
```

Comment:
Renamed all "defer" objects to "waitGroups" following Pierre's [suggestion](https://github.com/elastic/kibana/pull/154151#discussion_r1158453843) (less confusing).